### PR TITLE
feat(connectors): add MYOB AccountRight OAuth MCP connector

### PR DESCRIPTION
**Pull request by yiboyasss**
## Summary
- Adds a MYOB AccountRight OAuth MCP connector: contacts (customers/suppliers), sales invoices, purchase bills, and general ledger (accounts/tax codes) tools, following the granular `sme-*` data scope model.
- Extends the shared `generic_oauth_callback`/authorize-redirect builder in `auth.py` with the two pieces MYOB needs beyond the generic OAuth flow:
  - `prompt=consent` on the authorize redirect — MYOB only appends the `businessId` (the company file GUID, required for every API call) to the redirect when this is set, and there is no other way to discover it (the old company-file-listing endpoint no longer works under the granular-scope model).
  - Inline `user: {uid, username}` identity extraction from the token response, since MYOB has no dedicated userinfo endpoint.
- Verified the current MYOB auth model directly against the actively-maintained `uptick/pymyob` client's source (not just official docs, which are stale on this point): the old `x-myobapi-cftoken` company-file login is dead — the OAuth token plus `x-myobapi-key` is sufficient.

## Test plan
- [x] `pytest tests/web/tools/test_myob_mcp.py tests/web/test_myob_oauth.py tests/alembic/test_20260903_seed_myob_mcp_app.py` — new coverage, all passing
- [x] `pytest tests/web/test_generic_oauth_login.py tests/web/test_deputy_oauth.py tests/web/tools/test_salesforce_mcp.py tests/web/tools/test_deputy_mcp.py` — regression check on shared `auth.py` code paths, all passing
- [x] `ruff check` / `ruff format --check` / `mypy` on all new/changed files
- [x] Self-reviewed (two independent passes covering correctness and security/conventions); fixed dead exception-class plumbing and added output-size capping on single-record get/create/update tools whose `Lines` arrays are otherwise unbounded

### CODE REVIEW

File: src/xagent/web/api/auth.py [hunk 1]
```diff
@@ -2630,6 +2703,12 @@ def generic_oauth_callback(
                 )
                 or "longlife_refresh_token"
             )
+        # MYOB's own docs list `scope` as required here too, but the
```
Comment (rogercloud) [inline src/xagent/web/api/auth.py]: BLOCKING (Finding B): `scope` is never added for MYOB here (contrast with the `is_deputy` branch just above that adds it). The justification is that "the production-grade pymyob SDK's code-exchange request omits it and works," but `pymyob` is not a dependency of this repo (not in `pyproject.toml`, not vendored) — there's no checkable artifact here backing that claim. MYOB's own official developer docs list `scope` as a required field in the token-exchange POST body alongside `client_id, client_secret, code, redirect_uri, grant_type`. `tests/web/test_myob_oauth.py` mocks `requests.post` entirely and never asserts on the `data` payload, so CI passing here is not evidence the real endpoint accepts a scope-less exchange.

Suggest sending `data["scope"]` for MYOB the same way as the `is_deputy` branch above, or replacing this comment with a concrete, checkable citation (pinned pymyob source link/version, or a note this was manually verified against a live MYOB sandbox).

File: src/xagent/web/tools/mcp/myob.py [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+
+def _success(**payload: Any) -> str:
+    return json.dumps({"status": "success", **payload}, ensure_ascii=False)
+
+
+def _error(message: str) -> str:
+    return json.dumps({"status": "error", "message": message}, ensure_ascii=False)
+
+
+def _business_id() -> str:
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/myob.py:47]: Minor (non-blocking): `_business_id()` returns the raw `MYOB_BUSINESS_ID` env value without the `url_path_id` sanitizer already imported and used elsewhere in this file for `uid` params (e.g. dot-segment rejection, percent-encoding). Since `config.py`'s generic refresh handler can overwrite `instance_url` for any provider whose refresh response contains that key (gated only by presence/non-empty, not by `provider == "myob"`), applying `url_path_id(business_id, "business_id")` here would close a defense-in-depth gap and match the house pattern used for `uid`.

File: src/xagent/web/tools/mcp/myob.py [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+    concurrency via a ``RowVersion`` field embedded in the object body
+    itself (not an HTTP header) -- there is no partial-update verb. Rather
+    than requiring every caller to round-trip the object's other fields and
+    RowVersion themselves, this fetches the current body first and merges
+    the caller's changes on top of it, so a caller only has to name what's
+    actually changing.
+    """
+    safe_uid = url_path_id(uid, "uid")
+    current = _request("GET", f"{path}{safe_uid}/")
+    if not isinstance(current, dict):
```
Comment (rogercloud) [inline src/xagent/web/tools/mcp/myob.py]: Minor (non-blocking): `_request` returns `{}` on a 204/empty response, and `{}` passes `isinstance(current, dict)`, so `merged` becomes just the caller's fields — silently dropping `RowVersion` and any other existing fields from the PUT if MYOB ever returns an empty body for the pre-fetch GET. Suggest `if not current:` instead of `if not isinstance(current, dict):`. Also note: only `myob_update_customer` has a test today; `myob_update_sales_invoice`/`myob_update_purchase_bill`/`myob_update_supplier` have none.

**Review by rogercloud — CHANGES_REQUESTED** [review]
## Summary

This PR adds a MYOB AccountRight OAuth MCP connector: a new `myob.py` tool module (20 tools covering customers, invoices, purchase bills, suppliers, tax codes, and business info), an alembic migration seeding the MYOB provider/app rows (mirroring the existing Deputy seed migration pattern), and MYOB-specific extensions to the shared `generic_oauth_callback`/authorize-redirect builder in `auth.py` (a `prompt=consent` authorize param and inline `user:{uid,username}` identity extraction, since MYOB has no userinfo endpoint). The company-file `businessId` GUID is captured from the redirect and persisted into the existing `UserOAuth.instance_url` column, following the same pattern already used by Salesforce/Deputy.

**Blocking: yes** — 2 blocking findings (details below).

## Design-level assessment (Round 0)

Verdict: **acceptable with reservations**. Every new mechanism (the `prompt=consent` authorize tunable, inline user-identity extraction, reuse of `instance_url` for a provider-specific identifier) has 1-4 existing precedents in the same files (google/zoom/jira/linear for the authorize tunable; Salesforce for inline identity extraction; Salesforce/Deputy for `instance_url` reuse). The migration is a deliberate, near line-for-line copy of the existing Deputy seed migration, is genuinely needed (the registry-only seeder skips non-empty DBs), and is not redundant with the registry. `myob.py` follows the house connector skeleton faithfully; no new abstractions are invented. The approach does not stretch the shared abstraction out of shape — but there are real downstream gaps in how the captured `businessId` is subsequently validated/used, and in how the shared token-refresh path handles MYOB's response shape (see findings below).

## Findings

### BLOCKING

**A. MYOB token refresh crashes with `TypeError`; every connection breaks ~20 minutes after connect.**
`src/xagent/web/tools/config.py:1067-1069` (shared `refresh_oauth_token_if_needed`, pre-existing/unmodified by this PR, reached via the generic fallback path since MYOB has no dedicated branch unlike meta/zoom/deputy/linkedin/jira/salesforce):
```python
oauth_account.expires_at = datetime.now(timezone.utc) + timedelta(
    seconds=data["expires_in"]
)
```
MYOB's documented refresh response returns `expires_in` as a JSON **string** (e.g. `"1200"`), not an int. `timedelta(seconds="1200")` raises `TypeError`. The Meta branch at `config.py:899-902` explicitly does `int(data["expires_in"])`; the generic branch MYOB falls into does not. The exception is swallowed by a broad `except Exception` at `config.py:1084-1089` that logs only the exception type name and returns `False` — no user-visible error. The test fixture at `tests/web/test_myob_oauth.py:219` uses `"expires_in": 3600` as a plain int, which doesn't match MYOB's real string format, so the suite doesn't catch this.
This line is pre-existing, but this PR is what adds MYOB as the first caller that hits this generic path with a string `expires_in`, making a latent bug newly reachable and immediately production-impacting for every MYOB user (access tokens live 1200s = 20min; every MYOB connection silently stops refreshing shortly after being created).
Severity: HIGH. Blocking because it's a material unavailability of the feature this PR ships, hit in the standard/only usage path, with a small low-risk fix.
Fix: wrap in `int(data["expires_in"])` at `config.py:1069`, or add a MYOB-specific branch mirroring Meta's.

**B. `scope` is omitted from the MYOB token-exchange request based on an unverifiable claim, contradicting MYOB's own documented required params.**
`src/xagent/web/api/auth.py:2699-2711` (new code). There's an explicit `if is_deputy: data["scope"] = ...` branch just above, but no equivalent for MYOB — `scope` is simply never added by omission. The justifying comment states that "the production-grade pymyob SDK's code-exchange request omits it and works," but `pymyob` is not a project dependency (not in `pyproject.toml`, not vendored) — there's no checkable artifact backing the claim in this codebase. MYOB's own official developer docs list `client_id, client_secret, scope, code, redirect_uri, grant_type` as the required token-exchange POST body. Tests provide no coverage either way: `tests/web/test_myob_oauth.py` mocks `requests.post` entirely and never asserts on the `data` payload sent, so CI passing is not evidence the real MYOB endpoint accepts a scope-less exchange. The comment is also misplaced — it sits above a generic `headers = {...}` line shared by every provider, not next to any MYOB-specific branch.
Severity: Medium-High. Blocking because if MYOB does enforce `scope` as its docs state, every MYOB connect fails at the exchange step with zero CI signal, and the fix is cheap and safe.
Fix: send `data["scope"]` for MYOB the same way as Deputy, OR at minimum replace the unverifiable claim with a concrete, checkable citation (a pinned pymyob source link/version, or a note that this was manually verified against a live MYOB sandbox), and move the comment to sit with MYOB-specific code.

### Non-blocking (confirmed)

1. **`MYOB_API_KEY` (`static_env`) has no DB fallback; missing `example.env` entries.** `src/xagent/web/builtin_mcp_registry.py:1263-1286`. `client_id`/`client_secret` are DB-resolvable via `_resolve_oauth_secret`, but `static_env` is resolved purely from host env — an admin who configures MYOB only via the admin UI gets a successful OAuth connect followed by every tool call failing on a missing `MYOB_API_KEY`. This tradeoff is already self-documented in a comment at `builtin_mcp_registry.py:1268-1283` (calls it "a real deployment footgun"), but `example.env` has no MYOB entries (unlike SALESFORCE_*/DEPUTY_*/LINEAR_*). Suggest adding `MYOB_CLIENT_ID`/`MYOB_CLIENT_SECRET`/`MYOB_REDIRECT_URI` stub entries for operator discoverability.

2. **`businessId` is validated for GUID shape only, never for actual access, at connect time.** `src/xagent/web/api/auth.py:245-269` / `2426-2667`. A syntactically valid but wrong/stale/foreign `businessId` is persisted, connect reports success, and every subsequent tool call 401/403s (fail-closed, but not caught early). Unlike Deputy (which performs a live identity check at `auth.py:2992-3033`), MYOB has no live verification at callback time — though Salesforce is likewise presence/type-only, so this is at least consistent with one sibling. `myob_get_business_info` (`myob.py:255-268`) already makes the exact `GET {base}/{businessId}/` call that could close this gap if run during the callback.

3. **`_business_id()` skips the `url_path_id` sanitizer used everywhere else in this file; the shared refresh path can overwrite `instance_url` ungated by provider.** `src/xagent/web/tools/mcp/myob.py:47-51` interpolates the raw business id into the URL path without going through `url_path_id` (already imported and used 6x elsewhere in this file for `uid` params). Separately, `src/xagent/web/tools/config.py:1021-1050`'s generic refresh handler writes `oauth_account.instance_url = data["instance_url"]` for **any** provider whose refresh response has that key, gated only by presence/non-empty, unlike the Deputy write right below it which is gated by `normalized_provider == "deputy"`. Exploit path requires MYOB's own refresh response to carry a malicious `instance_url` (i.e. a trusted TLS-protected provider sending bad data), so this is defense-in-depth rather than a demonstrated exploit. Suggest applying `url_path_id(business_id, "business_id")` in `_business_id()` for house consistency.

4. **`_update_resource`'s empty-body guard doesn't catch `{}`, risking a silent field-wipe on PUT.** `src/xagent/web/tools/mcp/myob.py:231-251`. `_request` returns `{}` on a 204/empty response; `{}` passes `isinstance(current, dict)`, so `merged` becomes just the caller's fields, dropping `RowVersion` and all other existing fields from the PUT. Fix: `if not current:` instead of `if not isinstance(current, dict):`. Low real-world trigger likelihood since a GET 204'ing on a single existing MYOB record isn't documented/normal behavior, but worth tightening. Related: no test exists for `myob_update_sales_invoice`/`myob_update_purchase_bill`/`myob_update_supplier` (only `myob_update_customer` is tested) — add alongside the guard fix.

5. **One-connection-per-provider deletion model mismatches MYOB's 1:N company-file structure (pre-existing, not introduced by this PR).** `src/xagent/web/api/auth.py:3160-3168` calls `delete_scoped_user_oauth_accounts(...)`, which deletes every row for `(user_id, provider)` ignoring `provider_user_id`/`instance_url` before inserting a fresh row (confirmed byte-identical to base commit in `services/user_oauth.py:143-166`). Reconnecting to a different MYOB company file silently repoints every tool at different accounting data with no warning, and no test covers this. This is a pre-existing cross-provider limitation that MYOB is simply the first to expose as a real 1:N mismatch (vs 1:1 for Salesforce org / Deputy install). Worth a forward-looking note (e.g. reconnect warning or multi-connection model before MYOB GA) but not a regression from this PR.

6. **Test coverage gaps.** (a) `tests/web/test_myob_oauth.py:151-180` asserts `prompt=consent` on the authorize URL but never asserts the six `sme-*` scopes actually appear in it — the highest-risk untested behavior since `default_scopes` is empty and all scopes come from the app row. (b) Output-size capping is only tested for `myob_get_sales_invoice`, despite the PR description claiming capping on "get/create/update" — no create/update capping test exists. (c) No test for `_update_resource`'s empty-GET-response path (ties to #4). (d) `tests/alembic/test_20260903_seed_myob_mcp_app.py:299-307` `test_down_revision_matches_current_head` is tautological (compares two hardcoded literals, never queries `alembic heads`), so it can't actually catch a multiple-heads collision as its docstring claims.

7. **Minor reuse/consistency nits.** (a) `myob.py:170-193` `_success_with_capped_list` duplicates a simpler version of `salesforce.py`'s equivalent halving-loop pattern; also has an edge case where an all-empty-items payload can still exceed budget if `total_count`/`extra` alone are large (salesforce's page variant has a fallback for this). (b) `auth.py:2246-2284` now has 5 near-identical `if provider.lower() == "X": params["prompt"] = "consent"` blocks; a set-membership check would read better, though each carries a distinct justification comment worth preserving. (c) `myob_list_tax_codes` (`myob.py:591-603`) omits filter/orderby unlike all 5 sibling list tools, with no comment explaining why. (d) List responses expose `truncated`/`total_count` but no `has_more`/`next_skip` (salesforce's page variant has these). (e) `_create_resource`'s Location-header fallback (`myob.py:225-227`) doesn't URL-decode the trailing segment (low impact — MYOB IDs are GUIDs).

## Simplification opportunities

Minor, optional only: `myob.py`'s `_list_items` `isinstance(result, list)` branch (around line 152) handles a bare-array response shape no current call site produces; it's deliberately documented and has a dedicated test (`test_list_items_tolerates_bare_array`), so this is a low-value, optional trim rather than a defect — not requesting a change. The repeated `try/except`+`_error(str(e))` boilerplate across all 20 `@mcp.tool()` functions matches existing house style shared by every other MCP connector module (deputy.py, salesforce.py, hubspot.py, etc.), so it's not flagged as a PR-specific issue.

## Prior review status (gemini-code-assist[bot])

Both prior automated findings are verified fixed in the current code, no further action needed:
- Error-detail truncation now applies unconditionally regardless of source (`myob.py:107-114`).
- `provider_user_id` bool-type-confusion now explicitly excludes bool via `not isinstance(raw_provider_user_id, bool)` (`auth.py:3040-3047`).

## Blocking status & recommended decision

Recommended event: **REQUEST_CHANGES**

Blocking issues:
- [new] A: MYOB token refresh crashes with TypeError on the (string-typed) `expires_in` field, breaking every MYOB connection ~20 minutes after connect — `src/xagent/web/tools/config.py:1069`
- [new] B: `scope` is silently omitted from the MYOB token-exchange request on an unverifiable claim that contradicts MYOB's own documented required params — `src/xagent/web/api/auth.py:2699-2711`

**Comment by yiboyasss** [comment]
Both blocking findings (A: refresh crash on string `expires_in`, B: missing `scope` on the token exchange) verified against MYOB's own docs and fixed in c65699103, with regression tests against the documented string format and the required-scope example.

Non-blocking items 1, 3a, 4, 6a-c, 7c fixed in 85d2da324 (example.env entries, `url_path_id` on `_business_id()`, the empty-`{}` GET-response guard in `_update_resource`, missing update/capping test coverage, `myob_list_tax_codes` filter/orderby, and the tautological `down_revision` test now walking the real alembic graph).

Left as-is, deliberately:
- 2 (live businessId verification at connect time) — matches the existing Salesforce precedent already noted in the review.
- 3b (ungated `instance_url` overwrite on refresh) and 5 (one-connection-per-provider model) — both pre-existing shared-code behavior affecting every OAuth provider, not something this connector should special-case around; broader fixes there are out of scope for this PR.
- 7a (capped-list edge case) and 7d (has_more/next_skip) — no live call site currently triggers 7a, and 7d is a response-shape change across every list tool; both noted as optional/minor in the review, left for a follow-up if wanted.
- 7b (consolidating the 5 `prompt=consent` provider checks) and 7e (Location-header URL-decoding) — noted as low-value in the review itself; not applied to avoid losing the per-provider justification comments / no realistic GUID case affected.

**Review by rogercloud — APPROVED** [review]
## Summary

This PR adds a MYOB AccountRight OAuth MCP connector: 20 tools covering contacts (customers/suppliers), sales invoices, purchase bills, and general ledger (accounts/tax codes), following the granular `sme-*` scope model; an alembic migration seeding the MYOB provider/app rows (mirroring the existing Deputy seed pattern); and MYOB-specific extensions to the shared `generic_oauth_callback`/authorize-redirect builder in `auth.py` (`prompt=consent`, inline `user:{uid,username}` identity extraction, and a `businessId` GUID captured from the redirect and persisted into `UserOAuth.instance_url`). This is a re-review — the PR previously had `CHANGES_REQUESTED` from 2 blocking findings plus 3 non-blocking inline comments from a human reviewer, and 2 comments from the gemini-code-assist bot.

**Blocking: no — recommended event: APPROVE**

## Round 0 — Design assessment

Verdict: **acceptable with reservations**. Every new mechanism (the `prompt=consent` authorize tunable, inline user-identity extraction, reuse of `instance_url` for a provider-specific identifier) has existing precedent elsewhere in `auth.py` (other `prompt=consent` providers above it; Salesforce for inline identity extraction; Salesforce/Deputy for `instance_url` reuse of a non-Salesforce-shaped value). The migration is a deliberate near-line-for-line copy of the Deputy seed migration and is genuinely needed. `myob.py` follows the house connector skeleton faithfully. The approach does not stretch the shared abstraction out of shape, but there are a few loose ends worth tightening (below).

### Design-level findings

- **`src/xagent/web/api/auth.py:2297` and `:2441-2451` form an implicit, code-distance-separated contract.** The `businessId` extraction ~1500 lines away depends entirely on `prompt=consent` being set at the authorize step; nothing links the two beyond a comment. Consider a cross-reference comment at both ends, or consolidating MYOB's authorize-time and callback-time quirks into `oauth_provider_quirks.py` (which already exists and holds e.g. `requires_json_accept_header`).
- **`src/xagent/web/builtin_mcp_registry.py:1277-1289` / `src/xagent/web/tools/config.py:3615-3620`: `static_env` has no DB fallback.** An admin who configures MYOB only via the admin UI gets a successful OAuth connect followed by every tool call failing on a missing `MYOB_API_KEY` env var, and the failure is silent (no log) when the host env var is simply unset. This is now clearly documented (`example.env:769-782`, plus a registry comment), which meaningfully lowers the risk versus the earlier review pass, but the underlying gap (no DB-backed fallback path for `static_env`, unlike `client_id`/`client_secret` which do have one via `_resolve_oauth_secret`) is still there. Worth a follow-up.
- **`src/xagent/web/models/user_oauth.py:62-64`: the `instance_url` column comment is stale** — it still describes the column as "Salesforce (and no other provider here)" despite Deputy and now MYOB storing non-Salesforce-shaped values (MYOB stores a bare GUID, not a URL). Update the comment while touching this area.
- **`src/xagent/web/tools/mcp/myob.py:247-276` `_update_resource` does a GET-merge-PUT** with no optimistic-concurrency retry and PUTs back MYOB-computed/read-only fields (e.g. `URI`, computed totals) alongside the caller's changes — a classic lost-update race between the GET and the PUT. Not unique to this PR's design (other connectors PUT similarly), but worth a comment noting the tradeoff since MYOB's docs make a point of enforcing `RowVersion`-based optimistic concurrency.
- **`src/xagent/web/builtin_mcp_registry.py:1226`-area:** please double check the app's user-facing description accurately reflects that several tools are write-capable (create/update), not read-only lookups.

## Line-level findings

### Major

1. **`src/xagent/web/tools/mcp/myob.py:186-209`** (call sites `:312, :382, :453, :526, :597, :629`) — `_success_with_capped_list` halves items when the payload is over budget but never adjusts `total_count`/exposes a resumable offset (`next_skip`/`has_more`), so a caller doing `skip += top` pagination over a capped response will silently skip the halved-away records even though `truncated: true` is set. `src/xagent/web/tools/mcp/salesforce.py:127-160`'s `_success_with_capped_page` already solves exactly this. Blocking: no — the `truncated` flag does surface the condition to the caller/LLM, so this degrades usefulness rather than silently corrupting data, but it's worth aligning with the Salesforce helper.
2. **`src/xagent/web/builtin_mcp_registry.py:1277-1289` + `src/xagent/web/tools/config.py:3615-3620`** — `static_env` deployment footgun described above. Blocking: no — now documented in `example.env` and a registry comment, so an operator has a path to discover the requirement; still recommend a DB fallback in a follow-up.
3. **`src/xagent/web/api/auth.py:3072-3093`** — the MYOB identity branch silently leaves `provider_user_id`/`email` as `None` when `token_data["user"]` is absent or non-dict, with no `logger.warning`, unlike the Salesforce branch above it and the generic userinfo branch below it — this re-opens the NULL-`provider_user_id` uniqueness concern this same file guards against elsewhere. Blocking: no — requires the provider to omit/malform the `user` object in the token response, which is an edge case, but a one-line `logger.warning` here would match house style and aid debugging.
4. **`src/xagent/web/tools/mcp/myob.py:26,36` + `builtin_mcp_registry.py:1240-1247`** — the connector's load-bearing external-API assumptions (AccountRight base URL + `sme-*` granular scopes work together, `businessId` equals the redirect's company-file GUID, `x-myobapi-cftoken` is truly obsolete) can't be verified from this repo. Not a code defect, but worth a direct confirmation from the author that these were checked against a live MYOB sandbox (the PR description says the auth model was verified against `pymyob`'s source — please confirm the scope/base-URL combination was also live-tested, not just the two blocking auth issues that were fixed).

### Minor

5. **`tests/web/tools/test_myob_mcp.py`** — 6 of 20 tools (`myob_list_suppliers`, `myob_get_supplier`, `myob_create_supplier`, `myob_get_purchase_bill`, `myob_create_purchase_bill`, `myob_get_account`) are never invoked by any test, leaving their endpoint path strings unverified — a repointed `"Contact/Supplier/"` → `"Contact/Customer/"` typo would pass the full suite. Please add at least a smoke test per tool.
6. **`tests/web/test_myob_oauth.py:275`** — `assert oauth_account.expires_at is not None` doesn't pin the actual expiry duration; a `timedelta(seconds=...)` → `timedelta(minutes=...)` regression would pass all tests. Assert the actual delta.
7. **`src/xagent/web/api/auth.py:240-242`** — `_MYOB_BUSINESS_ID_PATTERN`'s trailing `$` anchor (the part that rejects a path-traversal suffix like `.../evil`) is untested; the existing parametrize covers short/empty/None/int but not a prefix match. `url_path_id` downstream still blocks the exploit, but the validator's own contract should be pinned by a test.
8. **`src/xagent/web/tools/mcp/myob.py:238`** — no test exercises the error-status path of any write tool (create/update); all error-path tests go through GET. Please add at least one non-2xx create/update test.
9. `src/xagent/web/tools/mcp/myob.py:168-183` — `_list_items`'s `total_count` for a bare-array response is just the page length, not the true collection total, so a caller could conclude there's no next page prematurely.
10. `src/xagent/web/tools/mcp/myob.py:186-209` is now the fourth near-identical copy of a capped-list helper across connectors (salesforce/deputy/chartmogul/myob) while `success_with_capped_dict` in `utils.py` is already shared — consider extracting `_success_with_capped_list` into `utils.py` too in a follow-up.
11. `src/xagent/web/tools/mcp/myob.py:297,368,437,510,583,616` — the `filter` parameter name shadows the `filter` builtin in all six list tools; consider `filter_expr` or similar.
12. `src/xagent/migrations/versions/20260903_seed_myob_mcp_app.py` `downgrade()` deletes the provider row without touching orphaned `user_oauth` rows for `myob` — a pre-existing pattern shared with sibling seed migrations, not introduced here, but worth fixing across the board eventually.
13. `src/xagent/web/api/auth.py:2297` — the new MYOB `prompt=consent` branch is byte-identical to an existing branch just above it for another provider; the growing per-provider `if provider.lower() == "X"` chain in `_generic_oauth_login` is due for a lookup-table refactor at some point (not blocking this PR).

## Simplification opportunities

- L240: `stdlib` Hand-rolled regex validates a GUID string that Python's `uuid` module already parses/validates. Replace with `try: uuid.UUID(business_id) except ValueError: return None`.
- L338: `delete` `_oauth_scope_str` is a single-caller wrapper (only the new MYOB token-exchange branch calls it) around the same two-line merge+join that `_generic_oauth_login` already inlines separately just above (`auth.py:2235-2236`). Inline the two lines at the MYOB call site instead of keeping a third variant of the same computation.

net: -6 lines possible

## Blocking status & recommended decision

**Blocking: no.**

The two previously blocking findings from the prior review round have both been verified fixed in the current HEAD (`496dc45c6`):
- [prior, now FIXED] Finding A — MYOB token refresh `TypeError` on string-typed `expires_in`: now wrapped in `int(...)` at `src/xagent/web/tools/config.py:1079` (and the pre-existing Meta branch at `:902` is consistent).
- [prior, now FIXED] Finding B — `scope` silently omitted from the MYOB token-exchange request: now sent via `_oauth_scope_str(db_provider.default_scopes, myob_app_scopes, provider)` at `src/xagent/web/api/auth.py:2745-2749`.

All 5 prior inline comments (2 from gemini-code-assist, 3 from the human reviewer) are also verified fixed in the current code:
- `myob.py:114` — error-detail truncation now applies unconditionally to `detail` regardless of source.
- `auth.py:3048` (now `:3084-3085`) — `provider_user_id` bool-type-confusion now explicitly excludes `bool` via `isinstance(..., (str, int)) and not isinstance(..., bool)`.
- `auth.py:2706` (scope) — see Finding B above.
- `myob.py:47` (now `:56`) — `_business_id()` now runs its value through `url_path_id(business_id, "MYOB_BUSINESS_ID")`.
- `myob.py:244` (now `:269`) — `_update_resource`'s guard is now `if not current or not isinstance(current, dict):`, correctly catching an empty-dict response.

No new confirmed finding from this pass independently meets the blocker criteria (incorrect user-visible result, data loss, security-boundary failure, or a severe availability/performance regression in a supported scenario) — the remaining majors are real but either self-mitigated (the `truncated` flag surfaces pagination gaps rather than silently hiding them) or require an edge-case provider response shape to trigger. Recommend addressing the open majors (pagination offset, `static_env` DB fallback, silent identity no-op) in this PR or a fast follow-up, but they do not block merge.

**Comment by yiboyasss** [comment]
Resolved the merge conflict against `main` (Employment Hero landed in the meantime — both `auth.py`'s identity-branch chain and the `myob`/`employment-hero` provider rows in `builtin_mcp_registry.py` needed manual reconciliation; also added a merge migration since the alembic graph had forked on `20260903_model_management`).

Fixed in 29a141cd1:
- Major #1 (pagination): `_success_with_capped_list` now returns `has_more`/`next_skip` reflecting what actually made it into the response post-halving, matching salesforce.py's own convention.
- Major #3 (silent identity no-op): added `logger.warning` for both a missing/malformed `user` object and a present-but-wrong-typed `uid`, matching the Salesforce branch's convention.
- Design: fixed the stale `instance_url` column comment, added a comment on `_update_resource`'s GET-then-PUT race, and reworded the app description to reflect its write-capable tools.
- Minor #5, #6, #7, #8: added smoke tests for the six previously-untested tools, a non-2xx write-tool error test, an actual expiry-duration assertion, and a test for the businessId regex's trailing anchor.
- Simplification L338: you're right that a single-caller wrapper wasn't actually shared with the authorize leg — reworked `_oauth_scope_str` into `_merged_oauth_scopes`, which now returns both the merged list and the joined string, and updated *both* legs to call it. This should resolve the tension between this suggestion and the earlier round's "these two legs will drift" concern at the same time; also pinned the authorize leg's actual `scope` query param in a test, closing that gap too.

On Major #4 (live-sandbox verification): to be fully transparent, I verified the auth model, base URL, and scope names against MYOB's official developer docs (fetched fresh, including worked examples for the token exchange body and response shapes) and against the actively-maintained `uptick/pymyob` reference client's actual source (not just its docs). I do not have a live MYOB developer/sandbox account, so none of this has been exercised against a real running MYOB business — only against documented behavior and a widely-used reference implementation's real behavior.

Left as-is, deliberately (not blocking, and each crosses into pre-existing shared code or an unrelated pattern rather than something this PR introduced):
- `static_env`'s missing DB fallback for `MYOB_API_KEY` — real, but fixing it means changing `tools/config.py`'s static_env resolution for every connector using it (currently also Google Ads), not just MYOB.
- Consolidating MYOB's authorize/callback quirks into `oauth_provider_quirks.py` — a reasonable reorganization, but touches how every provider's special-casing in `auth.py` is structured, not MYOB-specific.
- Extracting `_success_with_capped_list` into `utils.py` (now a 4th near-duplicate alongside salesforce/deputy/chartmogul) — agreed this is due for a shared helper, but that's a cross-file change touching three other connectors.
- Migration `downgrade()` not cleaning up orphaned `user_oauth` rows, and the growing per-provider `if provider.lower() == "X"` chain in `_generic_oauth_login` — both explicitly called out as pre-existing patterns shared across every seed migration / every provider, not introduced here.
- `_list_items`'s bare-array `total_count` fallback and the `filter` parameter name shadowing the builtin — genuine but zero live impact today (no endpoint currently returns a bare array; `filter` isn't used as a name inside those function bodies), left as noted rather than changed.
- Replacing the hand-rolled GUID regex with `uuid.UUID()` — considered this, but `uuid.UUID()` is more lenient than the regex (accepts non-hyphenated/braced forms) and its `str()` normalizes to lowercase, which would silently change the case of a stored businessId (MYOB's own GUIDs are typically uppercase) -- decided the current regex is the safer choice here, not just an oversight.

**Comment by yiboyasss** [comment]
Opened follow-up issues for the three items that came up repeatedly across review rounds and cross shared code beyond this PR:

- xorbitsai/xagent#2107 — static_env has no DB-backed fallback (Google Ads + MYOB affected today)
- xorbitsai/xagent#2108 — _success_with_capped_list duplicated across salesforce/deputy/chartmogul/myob
- xorbitsai/xagent#2109 — seed migration downgrade() leaves orphaned user_oauth rows (shared by every builtin connector's migration)
